Common/Securities/Futures/FuturesExpiryFunctions.cs

```diff
@@ -40,44 +39,487 @@ public class FuturesExpiryFunctions
+        /// </summary>
+        private static DateTime BusinessDayPrecedingThirdWednesdayAtNineSixteen(DateTime time)
+        {
+            // Get Third Wednesday
+            var daysInMonth = DateTime.DaysInMonth(time.Year, time.Month);
+            var ThirdWednesday = (from day in Enumerable.Range(1, daysInMonth)
+                                  where new DateTime(time.Year, time.Month, day).DayOfWeek == DayOfWeek.Wednesday
+                                  select new DateTime(time.Year, time.Month, day)).ElementAt(2);
+
+            // Now get the second business day preceding third Wednesday
+            var previousDay = NthBusinessDay(ThirdWednesday, -1);
+            return new DateTime(previousDay.Year, previousDay.Month, previousDay.Day, 14, 16, 0);
+        }
+
+        /// <summary>
+        /// Mehtod to retrieve 12:01 on last business day of the calendar month.
+        /// </summary>
+        private static DateTime LastBusinessDay(DateTime time, TimeSpan t)
+        {
+            var daysInMonth = DateTime.DaysInMonth(time.Year, time.Month);
+            return (from day in Enumerable.Range(1, daysInMonth)
+                    let _day = new DateTime(time.Year, time.Month, day, t.Hours, t.Minutes, t.Seconds)
+                    where notHoliday(_day)
+                    select _day).Reverse().ElementAt(0);
+        }
+
+        /// <summary>
+        /// Method to retrieve 12:01 pm on the seventh business day preceding the last business day of the delivery month.
+        /// </summary>
+        private static DateTime SeventhBusinessDayPreceedingLastBusinessDay(DateTime time)
+        {
+            var daysInMonth = DateTime.DaysInMonth(time.Year, time.Month);
+            var lastBusinessDay = (from day in Enumerable.Range(1, daysInMonth)
+                                   let _day = new DateTime(time.Year, time.Month, day)
+                                   where notHoliday(_day)
+                                   select _day).Reverse().ElementAt(0);
+            var seventhPreceding = NthBusinessDay(lastBusinessDay, -7);
+            return new DateTime(seventhPreceding.Year, seventhPreceding.Month, seventhPreceding.Day, 12, 01, 0);
+        }
+
+        /// <summary>
+        /// Method to retrieve last business day of the month preceding the delivery month. 
+        /// </summary>
+        private static DateTime LastBusinessDayPrecedingMonth(DateTime time)
+        {
+            var precedingMonth = time.AddMonths(-1);
+            var daysInMonth = DateTime.DaysInMonth(precedingMonth.Year, precedingMonth.Month);
+            var lastBusinessDay = (from day in Enumerable.Range(1, daysInMonth)
+                                   let _day = new DateTime(precedingMonth.Year, precedingMonth.Month, day)
+                                   where notHoliday(_day)
+                                   select _day).Reverse().ElementAt(0);
```

Comment:
Local variables, even within lambdas, typically do not use the underscore in their name